// ng2-ckeditor/lib/bundles/ng2-ckeditor.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/forms')) :
    typeof define === 'function' && define.amd ? define('ng2-ckeditor', ['exports', '@angular/core', '@angular/common', '@angular/forms'], factory) :
    (global = global || self, factory(global['ng2-ckeditor'] = {}, global.ng.core, global.ng.common, global.ng.forms));
}(this, function (exports, core, common, forms) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    /**
     * CKGroup component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
     *      <ckbutton [name]="'SaveButton'" [command]="'saveCommand'" (click)="save($event)"
     *                [icon]="'/save.png'" [toolbar]="'customGroup,1'" [label]="'Save'">
     *      </ckbutton>
     *   </ckeditor>
     */
    var CKButtonDirective = /** @class */ (function () {
        function CKButtonDirective() {
            this.click = new core.EventEmitter();
        }
        CKButtonDirective.prototype.initialize = function (editor) {
            var _this = this;
            editor.instance.addCommand(this.command, {
                exec: function (evt) {
                    _this.click.emit(evt);
                },
            });
            editor.instance.ui.addButton(this.name, {
                label: this.label,
                command: this.command,
                toolbar: this.toolbar,
                icon: this.icon,
            });
        };
        CKButtonDirective.prototype.ngOnInit = function () {
            if (!this.name)
                throw new Error('Attribute "name" is required on <ckbutton>');
            if (!this.command)
                throw new Error('Attribute "command" is required on <ckbutton>');
        };
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKButtonDirective.prototype, "click", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "label", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "command", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "toolbar", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "name", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKButtonDirective.prototype, "icon", void 0);
        CKButtonDirective = __decorate([
            core.Directive({
                selector: 'ckbutton',
            })
        ], CKButtonDirective);
        return CKButtonDirective;
    }());

    /**
     * CKGroup component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500">
     *      <ckgroup [name]="'exampleGroup2'" [previous]="'1'" [subgroupOf]="'exampleGroup1'">
     *          .
     *          .
     *      </ckgroup>
     *   </ckeditor>
     */
    var CKGroupDirective = /** @class */ (function () {
        function CKGroupDirective() {
        }
        CKGroupDirective.prototype.ngAfterContentInit = function () {
            var _this = this;
            // Reconfigure each button's toolbar property within ckgroup to hold its parent's name
            this.toolbarButtons.forEach(function (button) { return (button.toolbar = _this.name); });
        };
        CKGroupDirective.prototype.initialize = function (editor) {
            editor.instance.ui.addToolbarGroup(this.name, this.previous, this.subgroupOf);
            // Initialize each button within ckgroup
            this.toolbarButtons.forEach(function (button) {
                button.initialize(editor);
            });
        };
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKGroupDirective.prototype, "name", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], CKGroupDirective.prototype, "previous", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKGroupDirective.prototype, "subgroupOf", void 0);
        __decorate([
            core.ContentChildren(CKButtonDirective),
            __metadata("design:type", core.QueryList)
        ], CKGroupDirective.prototype, "toolbarButtons", void 0);
        CKGroupDirective = __decorate([
            core.Directive({
                selector: 'ckgroup',
            })
        ], CKGroupDirective);
        return CKGroupDirective;
    }());

    /**
     * CKEditor component
     * Usage :
     *  <ckeditor [(ngModel)]="data" [config]="{...}" debounce="500"></ckeditor>
     */
    var CKEditorComponent = /** @class */ (function () {
        /**
         * Constructor
         */
        function CKEditorComponent(zone) {
            this.zone = zone;
            this.change = new core.EventEmitter();
            this.editorChange = new core.EventEmitter();
            this.ready = new core.EventEmitter();
            this.blur = new core.EventEmitter();
            this.focus = new core.EventEmitter();
            this.contentDom = new core.EventEmitter();
            this.fileUploadRequest = new core.EventEmitter();
            this.fileUploadResponse = new core.EventEmitter();
            this.paste = new core.EventEmitter();
            this.drop = new core.EventEmitter();
            this._value = '';
        }
        CKEditorComponent_1 = CKEditorComponent;
        Object.defineProperty(CKEditorComponent.prototype, "value", {
            get: function () {
                return this._value;
            },
            set: function (v) {
                if (v !== this._value) {
                    this._value = v;
                    this.onChange(v);
                }
            },
            enumerable: true,
            configurable: true
        });
        CKEditorComponent.prototype.ngOnChanges = function (changes) {
            if (changes.readonly && this.instance) {
                this.instance.setReadOnly(changes.readonly.currentValue);
            }
        };
        /**
         * On component destroy
         */
        CKEditorComponent.prototype.ngOnDestroy = function () {
            var _this = this;
            if (this.instance) {
                setTimeout(function () {
                    if (_this.instance) {
                        _this.instance.removeAllListeners();
                        CKEDITOR.instances[_this.instance.name].destroy();
                        _this.instance.destroy();
                        _this.instance = null;
                    }
                });
            }
        };
        /**
         * On component view init
         */
        CKEditorComponent.prototype.ngAfterViewInit = function () {
            this.ckeditorInit(this.config || {});
        };
        /**
         * On component view checked
         */
        CKEditorComponent.prototype.ngAfterViewChecked = function () {
            this.ckeditorInit(this.config || {});
        };
        /**
         * Value update process
         */
        CKEditorComponent.prototype.updateValue = function (value) {
            var _this = this;
            this.zone.run(function () {
                _this.value = value;
                _this.onChange(value);
                _this.onTouched();
                _this.change.emit(value);
            });
        };
        /**
         * CKEditor init
         */
        CKEditorComponent.prototype.ckeditorInit = function (config) {
            var _this = this;
            if (typeof CKEDITOR === 'undefined') {
                console.warn('CKEditor 4.x is missing (http://ckeditor.com/)');
            }
            else {
                // Check textarea exists
                if (this.instance || !this.documentContains(this.host.nativeElement)) {
                    return;
                }
                if (this.readonly) {
                    config.readOnly = this.readonly;
                }
                // CKEditor replace textarea
                this.instance = CKEDITOR.replace(this.host.nativeElement, config);
                // Set initial value
                this.instance.setData(this.value);
                // listen for instanceReady event
                this.instance.on('instanceReady', function (evt) {
                    // if value has changed while instance loading
                    // update instance with current component value
                    if (_this.instance.getData() !== _this.value) {
                        _this.instance.setData(_this.value);
                    }
                    // send the evt to the EventEmitter
                    _this.ready.emit(evt);
                });
                // CKEditor change event
                this.instance.on('change', function (evt) {
                    _this.onTouched();
                    var value = _this.instance.getData();
                    if (_this.value !== value) {
                        // Debounce update
                        if (_this.debounce) {
                            if (_this.debounceTimeout)
                                clearTimeout(_this.debounceTimeout);
                            _this.debounceTimeout = setTimeout(function () {
                                _this.updateValue(value);
                                _this.debounceTimeout = null;
                            }, parseInt(_this.debounce));
                            // Live update
                        }
                        else {
                            _this.updateValue(value);
                        }
                    }
                    // Original ckeditor event dispatch
                    _this.editorChange.emit(evt);
                });
                // CKEditor blur event
                this.instance.on('blur', function (evt) {
                    _this.blur.emit(evt);
                });
                // CKEditor focus event
                this.instance.on('focus', function (evt) {
                    _this.focus.emit(evt);
                });
                // CKEditor contentDom event
                this.instance.on('contentDom', function (evt) {
                    _this.contentDom.emit(evt);
                });
                // CKEditor fileUploadRequest event
                this.instance.on('fileUploadRequest', function (evt) {
                    _this.fileUploadRequest.emit(evt);
                });
                // CKEditor fileUploadResponse event
                this.instance.on('fileUploadResponse', function (evt) {
                    _this.fileUploadResponse.emit(evt);
                });
                // CKEditor paste event
                this.instance.on('paste', function (evt) {
                    _this.paste.emit(evt);
                });
                // CKEditor drop event
                this.instance.on('drop', function (evt) {
                    _this.drop.emit(evt);
                });
                // Add Toolbar Groups to Editor. This will also add Buttons within groups.
                this.toolbarGroups.forEach(function (group) {
                    group.initialize(_this);
                });
                // Add Toolbar Buttons to Editor.
                this.toolbarButtons.forEach(function (button) {
                    button.initialize(_this);
                });
            }
        };
        /**
         * Implements ControlValueAccessor
         */
        CKEditorComponent.prototype.writeValue = function (value) {
            this._value = value;
            if (this.instance)
                this.instance.setData(value);
        };
        CKEditorComponent.prototype.onChange = function (_) { };
        CKEditorComponent.prototype.onTouched = function () { };
        CKEditorComponent.prototype.registerOnChange = function (fn) {
            this.onChange = fn;
        };
        CKEditorComponent.prototype.registerOnTouched = function (fn) {
            this.onTouched = fn;
        };
        CKEditorComponent.prototype.documentContains = function (node) {
            return document.contains ? document.contains(node) : document.body.contains(node);
        };
        var CKEditorComponent_1;
        __decorate([
            core.Input(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "config", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Boolean)
        ], CKEditorComponent.prototype, "readonly", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", String)
        ], CKEditorComponent.prototype, "debounce", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "change", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "editorChange", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "ready", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "blur", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "focus", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "contentDom", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "fileUploadRequest", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "fileUploadResponse", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "paste", void 0);
        __decorate([
            core.Output(),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "drop", void 0);
        __decorate([
            core.ViewChild('host'),
            __metadata("design:type", Object)
        ], CKEditorComponent.prototype, "host", void 0);
        __decorate([
            core.ContentChildren(CKButtonDirective),
            __metadata("design:type", core.QueryList)
        ], CKEditorComponent.prototype, "toolbarButtons", void 0);
        __decorate([
            core.ContentChildren(CKGroupDirective),
            __metadata("design:type", core.QueryList)
        ], CKEditorComponent.prototype, "toolbarGroups", void 0);
        __decorate([
            core.Input(),
            __metadata("design:type", Object),
            __metadata("design:paramtypes", [Object])
        ], CKEditorComponent.prototype, "value", null);
        CKEditorComponent = CKEditorComponent_1 = __decorate([
            core.Component({
                selector: 'ckeditor',
                providers: [
                    {
                        provide: forms.NG_VALUE_ACCESSOR,
                        useExisting: core.forwardRef(function () { return CKEditorComponent_1; }),
                        multi: true,
                    },
                ],
                template: "\n    <textarea #host></textarea>\n  "
            }),
            __metadata("design:paramtypes", [core.NgZone])
        ], CKEditorComponent);
        return CKEditorComponent;
    }());

    /**
     * CKEditorModule
     */
    var CKEditorModule = /** @class */ (function () {
        function CKEditorModule() {
        }
        CKEditorModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule],
                declarations: [CKEditorComponent, CKButtonDirective, CKGroupDirective],
                exports: [CKEditorComponent, CKButtonDirective, CKGroupDirective],
            })
        ], CKEditorModule);
        return CKEditorModule;
    }());

    exports.CKEditorModule = CKEditorModule;
    exports.ɵa = CKEditorComponent;
    exports.ɵb = CKButtonDirective;
    exports.ɵc = CKGroupDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
